### DISCUSSION TITLE
Disable clipboard URL by default

### DISCUSSION THREAD

**Discussion by D10f** [philc/vimium#4592] [General]
The title says it all, I think this feature is unnecessary and should be disabled by default to prevent accidental data being sent to Google.

Now that I know this setting exists, I can disable it, but being the first time using this plugin it's quite hard to believe that my entire clipboard contents gets send automatically to Google if I accidentally press `P`.

Why does this even exist? I could understand it if it used the selection under the cursor, maybe to search for word definitions or tings like that. But the clipboard? That could potentially contain sensitive information. And why not just use my default search engine? At the very least, that's what it should search with.

The same functionality can be achieved just as easily with default keyboard shortcuts like <kbd>Ctrl</kbd>+<kbd>L</kbd>, <kbd>Ctrl</kbd>+<kbd>J</kbd> or <kbd>Ctrl</kbd>+<kbd>K</kbd>, depending on your setup, and then paste the contents of the clipboard. There's no need for this to be enabled by default. Those who really want this feature can enable it easily once they know about it, but the other way around exposes people to accidentally leak potentially sensitive data to Google.

**Comment by philg-dev** [comment]
## Regarding default keybindings

Just straight out saying "this feature is unnecessary" seems a bit drastic. I'm sure there's plenty of Vimium users that use it very regularly. Vimium is most likely an addon that's used almost exclusively by power users, some of which might just like interactions that save a tiny amount of clicks / keystrokes.

While I do not necessarily agree with straight up disabling the `p` and `P` shortcuts by default, I do agree that a bunch of the default functionality of Vimium could be a little bit frustrating to new users, before they get used to using Vimium. Especially folks that aren't used to any other "vim-like" user interaction.

I think it might be nice to at least have some FAQ entry, a beginner's guide or a small section in the Readme with a selection of unmaps for potentially "annoying" keybindings for beginners.

When I started out, I unmapped the following keybindings, just to avoid any "accidents" (esp. affecting the current tab):

```
# x closes tab
unmap x
# r reloads tab
unmap r
# p opens clipboard URL in current tab (Vimium default search engine, if not a link)
unmap p
# P opens clipboard URL in a new tab (Vimium default search engine, if not a link)
unmap P
```

I've been using Vimium for a few months now and I've only re-activated the `P` default binding because it's convenient and even if I were to accidentally trigger it somehow, it would just be a in a new tab, which I can close quickly again. For `x` and `r` I personally prefer the browser native keybindings (`F5` / `CTRL+R` to reload, `CTRL+W` to close current tab).

## Regarding the browser's default engine:

From what I've seen, this has been handled in #4439 
I'm not exactly sure what the state of this is, in the meantime. It seems to be merged for quite a long time, but even on a fresh install on Firefox it doesn't seem to be working yet(?) Maybe it's not part of an actual release on the AMO and Google Extension Store yet. Maybe the additional add-on permission is an issue in that process - I don't know. 

As for now, you can just change Vimium's default search engine. That way clipboard content, that isn't a link, will open in the search engine of your choice. Currently if you just set Vimium's default search engine to an empty string, Vimium will not do anything when encountering non-URL text, so that might be good for you. As a heads-up though: From what I understand, an empty string might lead to browser default search engine in the future, when the mentioned pull request is part of a release.

**Reply by D10f** [reply]
Hi, thank you for the input.

Maybe saying it's unnecessary was indeed a bit too much, but I still find this feature out of place as a default option. The keyword that I want to emphasize being "default". I can see how this may be appealing to some, I'm not arguing against its usefulness, just the design decision to include it by default and to behave in such a way.

I agree that it's unreasonable to expect a 1:1 feature comparison between a text editor and a browser &mdash; it makes sense that certain keybindings are repurposed. However, <kbd>P</kbd> could've just as easily been assigned to open a new tab into "about:**p**references", for example, which is also something that I think most people frequently do. Granted, not as often as searching for stuff online, but it's still reasonable, easy to remember and doesn't put the user's data at risk.

For context: I found out about this while I was logging into a site using the Auto-Type feature of KeePassXC (password manager). The site I was logging into had just changed layouts, and so the auto-type sequence pressed <kbd>TAB</kbd> one too many times, taking focus away from the input field, making the Vimium keybindings work in normal mode. As it just so happened, one of the characters in my password was <kbd>P</kbd>, triggering the online search immediately with my clipboard in it.

Reading from the clipboard is not something that applications should do carelessly. When in doubt, be explicit by requiring user action or, at the very least, inform the user about it. A good example of this put into practice can be found in mobile web browsers: when focusing the address bar, if they detect that your clipboard contains a URL, they'll offer to visit that. The user is informed and given the opportunity to react, and then again only when it's an URL and not whatever random data is there.

To be a little more constructive about this let me suggest that the search engine string is set to an empty string by default. As it stands right now, Vimium doesn't read the default search engine anyway, so even if I wanted to use this feature but with another search engine I would have to go into the settings to change it.

By prompting the user to enable the feature it can make an informed decision as to whether it's worth it or not, and how should it behave.

Anyway, hope this can be taken into consideration for future releases. Happy Holidays!

**Reply by philg-dev** [reply]
I've actually encountered almost the same issue recently with the KeePassXC auto-type scenario.

I just had the idea that it would potentially be a nice feature to have an option to dynamically disable a set of Vimium functions, when Vimium detects a **password field on a website**. I know there's still other potential scenarios might be potentially concerning to you... But that would definitely be something that could be implemented fairly easily. Might be a reasonable compromise.

Regarding default settings, it's always a balancing act between making the out-of-the-box experience the most comfortable vs. having funcitonality as "opt-in" that might be annoying / upsetting for an unknown amount of users. Maybe the author finds this concerning enough to change it. I personally don't mind it the way it is.

**Reply by D10f** [reply]
Yes, defaults are tough... what one wants destroys the user experience of the next. I opened this under discussion to see if there would be enough interest or at least to bring up the issue (without cluttering the actual issue tracker).
